server/routes/archiveGoal/archiveGoalController.test.ts

```diff
@@ -0,0 +1,258 @@
+        req as undefined as Request,
+        res as undefined as Response,
+        next as undefined as NextFunction,
+      )
+
+      // Then
+      expect(next).toHaveBeenCalledWith(expectedError)
+      expect(req.session.archiveGoalForm).toBeUndefined()
+    })
+
+    it('should not get archive goal view given requested goal reference is not part of the prisoners action plan', async () => {
+      // Given
+      const someOtherGoalReference = 'd31d22bc-b9be-4d13-9e47-d633d6815454'
+      const goal = aValidGoal({ goalReference: someOtherGoalReference })
+      const actionPlan = aValidActionPlanWithOneGoal({ prisonNumber, goal })
+      educationAndWorkPlanService.getActionPlan.mockResolvedValue(actionPlan)
+
+      const expectedError = createError(404, `Goal ${goalReference} does not exist in the prisoner's plan`)
+
+      // When
+      await controller.getArchiveGoalView(
+        req as undefined as Request,
+        res as undefined as Response,
+        next as undefined as NextFunction,
+      )
+
+      // Then
+      expect(next).toHaveBeenCalledWith(expectedError)
+      expect(req.session.archiveGoalForm).toBeUndefined()
+    })
+  })
+  describe('submitArchiveGoalForm', () => {
+    it('should redirect to review archive goal page given action is submit-form and validation passes', async () => {
+      // Given
+      req.params.prisonNumber = 'A1234GC'
+
+      const archiveGoalForm = aValidArchiveGoalForm(goalReference)
+      req.body = { ...archiveGoalForm }
+
+      mockedValidateArchiveGoalForm.mockReturnValue([])
+
+      // When
+      await controller.submitArchiveGoalForm(
+        req as undefined as Request,
+        res as undefined as Response,
+        next as undefined as NextFunction,
+      )
+
+      // Then
+      expect(res.redirect).toHaveBeenCalledWith(`/plan/${prisonNumber}/goals/${goalReference}/archive/review`)
+      expect(mockedValidateArchiveGoalForm).toHaveBeenCalledWith(archiveGoalForm)
+    })
+    it('should redirect to archive goal form given validation fails', async () => {
+      // Given
+      req.params.prisonNumber = 'A1234GC'
+      req.body = { reference: goalReference, title: 'Learn German' }
+
+      errors = [{ href: '#reason', text: 'bang!' }]
+      mockedValidateArchiveGoalForm.mockReturnValue(errors)
+
+      const expectedArchiveGoalForm: ArchiveGoalForm = { reference: goalReference, title: 'Learn German' }
+
+      // When
+      await controller.submitArchiveGoalForm(
+        req as undefined as Request,
+        res as undefined as Response,
+        next as undefined as NextFunction,
+      )
+
+      // Then
+      expect(res.redirectWithErrors).toHaveBeenCalledWith(
+        `/plan/${prisonNumber}/goals/${goalReference}/archive`,
+        errors,
+      )
+      expect(req.session.archiveGoalForm).toEqual(expectedArchiveGoalForm)
+    })
+  })
+  describe('getReviewArchiveGoalView', () => {
+    it('should get the view with the form', async () => {
```

Comment:
This test is based around the `ArchiveGoalForm` being on the session (ie. the previous page left it there)

Do we also need a test for when the form is not on the session? (ie. the user has book marked the page / arrived on the page unexpectedly?) In this case I think I'd expect the user to be redirected back to the main Archive Goal page (ie. the page that logically comes before this one)